First we have to import libraries such as the <em>pandas</em> and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

Dataset provided in the assignment are:

The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:
1. Cement
2. Blast Furnace Slag
3. Fly Ash
4. Water
5. Superplasticizer
6. Coarse Aggregate
7. Fine Aggregate

To read the dataset into a <em>pandas</em> dataframe:

In [2]:
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So, the data suggests that the first sample has 540 cubic metres of cement, 0 cubic metres of blast furnace slag, 0 cubic metre of fly ash, 162 cubic meter of water, 2.5 cubic metres of superplaticizer, 1040 cubic metres of coarse aggregate and, 676 cubic metres of fine aggregate. 
This concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

<h4>So to check the number of data points:</h4>

In [3]:
concrete_data.shape

(1030, 9)

So, there are total of 1030 samples to train our model on. So, we should be very careful not to overfit the training data.

We can use <i><u>concrete_data.describe()</u></i> to check the dataset for any missing values.

In [4]:
concrete_data.describe()
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

So we can colclude that this dataset is very clean and thus it can be used safely to build our model.

<h4>Split data into predictors and target</h4>

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [5]:
data_columns = concrete_data.columns
predictors = concrete_data[data_columns[data_columns != 'Strength']] # to get all columns except Strength
target = concrete_data['Strength'] # to get Strength column

To make a sanity check  forthe data again for sake of the predictors and the target dataframes.

In [6]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [7]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

So, it looks clean again.

In [8]:
number_cols = predictors.shape[1] # number of predictors
number_cols

8

<a id="item1"></a>

<a id="item1"></a>

## Importing Keras

#### Now we have to import the Keras library for the actual work

In [9]:
import keras

To use Keras, we first have to install the TensorFlow backend, which would be used to install the Keras library.

Now, to import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [10]:
from keras.models import Sequential
from keras.layers import Dense

In [11]:
def regression_model():
    # creating the model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(number_cols,)))
    model.add(Dense(1))
    
    # compiling the model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The regression_model function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [12]:
from sklearn.model_selection import train_test_split

Now, let's split the data into a training and test sets by holding 30% of the data for testing

In [13]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

### Training and Testing the Network

Call the function now to create our model.

In [14]:
# building the model
model = regression_model()

Next, we will train the model for 50 epochs.


In [15]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
23/23 [==============================] - 0s 1ms/step - loss: 140486.3281
Epoch 2/50
23/23 [==============================] - 0s 910us/step - loss: 52761.3438
Epoch 3/50
23/23 [==============================] - 0s 764us/step - loss: 15300.5420
Epoch 4/50
23/23 [==============================] - 0s 767us/step - loss: 4078.8591
Epoch 5/50
23/23 [==============================] - 0s 741us/step - loss: 1904.8817
Epoch 6/50
23/23 [==============================] - 0s 810us/step - loss: 1643.3110
Epoch 7/50
23/23 [==============================] - 0s 812us/step - loss: 1587.2657
Epoch 8/50
23/23 [==============================] - 0s 767us/step - loss: 1541.0066
Epoch 9/50
23/23 [==============================] - 0s 810us/step - loss: 1492.3256
Epoch 10/50
23/23 [==============================] - 0s 861us/step - loss: 1445.9077
Epoch 11/50
23/23 [==============================] - 0s 1ms/step - loss: 1398.6022
Epoch 12/50
23/23 [==============================] - 0s 824us/step - loss:

Next we need to evaluate the model on the test data.

In [16]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 885us/step


595.7080078125

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [17]:
from sklearn.metrics import mean_squared_error

In [18]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

595.7080658078809 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [20]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 44.85261917114258
10/10 [==============================] - 0s 665us/step
MSE 2: 48.12493896484375
10/10 [==============================] - 0s 733us/step
MSE 3: 38.66611862182617
10/10 [==============================] - 0s 776us/step
MSE 4: 47.988182067871094
10/10 [==============================] - 0s 808us/step
MSE 5: 42.89168167114258
10/10 [==============================] - 0s 647us/step
MSE 6: 45.06026840209961
10/10 [==============================] - 0s 700us/step
MSE 7: 53.152923583984375
10/10 [==============================] - 0s 770us/step
MSE 8: 46.89166259765625
10/10 [==============================] - 0s 660us/step
MSE 9: 40.754676818847656
10/10 [==============================] - 0s 670us/step
MSE 10: 50.810279846191406
10/10 [==============================] - 0s 792us/step
MSE 11: 43.492122650146484
10/10 [==============================] - 0s 717us/step
MSE 12: 41.215057373046875
10/10 [==============================] - 0s 728us/step
MSE 13: 51.967262268066406
10/1